### PR TITLE
Export request to CURL format

### PR DESCRIPTION
**Pull request by AlphaTechnolog**
The idea is to implement #63 but in a way that it's open for also implementing #64 since those could be placed in the same panel but with more options in the combobox.

> [!NOTE]
> Still this is work in progress and isn't fully done, just leaving this here as a draft to document a little bit the process 😅 

image

> The icons aren't supposed to look that way, it's just because of some weird environment im coding in XD

**Status**: Ready for review.

### CODE REVIEW

**Comment by AlphaTechnolog** [comment]
Also recommendations on code quality and others are really really welcome, i've been trying my best to read all the code i've could to be able to do this because i'm basically rust + gtk noob lmao

**Comment by AlphaTechnolog** [comment]
Right now this is only working with raw body format, but im not really sure how to export body for others formats such as `Urlencoded` and `Multipart`. How would you handle this @danirod ? For the raw body am just pasting it around `format!("-d '{body}'")` and skipping it for the missing body types mentioned before.

**Comment by AlphaTechnolog** [comment]
Another thing is that i added now a copy request contents button, the thing is that i'm not really sure now how could i make it display a little toast like the errors ones but when the user copies the thing... I saw the method error_toast() in window or something similar, but i can't seem to find a way to obtain the window instance from within the code.rs widget, is there any way to fetch it? or at least fetch the toaster widget which is wrapping the app?

After this and after discussing about the handling of others formats of body, i think this pr should be ready to review

**Comment by danirod** [comment]
Sorry, I am slow to respond. I still need to read the code, but let me say that the screenshot looks incredible and I am in love with how modern the Copy button looks. I agree that the panel could be reused to provide code generation for other programming languages. It reminds me of what regex101 does in their Export dialog. One dialog, many languages.

Let me try to provide information here for what you're asking.

> How would you handle this @danirod ? For the raw body am just pasting it around format!("-d '{body}'") and skipping it for the missing body types mentioned before.

That's a very good question that I didn't consider until now.

I've just made a test using Firefox to see how do they implement "Copy as cURL" when there is a payload, and it seems that all they do is to just convert to a raw body, similar to when it does a request, and use -d parameters.

After all, a urlencoded body like `name: john` and `surname`: `smith` can be converted to `-d 'name=john&surname=smith'`.

However, it seems that if one of the fields is binary (for instance, files), they don't copy the file to the curl command. I suppose that they don't want random bytes in the clipboard.


> i can't seem to find a way to obtain the window instance from within the code.rs widget

There is a method in the Gtk.Application class called `get_active_window()`. Using `CarteroApplication::get()`, you should be able to retrieve the CarteroApplication instance and then it should let you use the `get_active_window()` method.

I am looking for examples and there is one in app.rs here, but a cast is required from GtkWindow to CarteroWindow.

https://github.com/danirod/cartero/blob/74c560081fbd67af442c6fbdb7dfb98687b22fde/src/app.rs#L67-L70

To be fair, it's true that the `toast_error` assumes that all you have is errors. Don't be afraid to add a similar method that just accepts a string to toast it if you need it. The `toast_error` method is currently rather limited.

**Comment by AlphaTechnolog** [comment]
Hey! 👋🏻 

> Sorry, I am slow to respond. I still need to read the code, but let me say that the screenshot looks incredible and I am in love with how modern the Copy button looks.

Don't worry! it's ok :) also thank you, yeah i felt like it would be a great addition since all one wants to do is click one button to get instantly the code :)

> I agree that the panel could be reused to provide code generation for other programming languages. It reminds me of what regex101 does in their Export dialog. One dialog, many languages.

Yeah, it would be easier for the user to just do it one single view or panel, 100% agree

> That's a very good question that I didn't consider until now.

> I've just made a test using Firefox to see how do they implement "Copy as cURL" when there is a payload, and it seems that all they do is to just convert to a raw body, similar to when it does a request, and use -d parameters.

> After all, a urlencoded body like name: john and surname: smith can be converted to -d 'name=john&surname=smith'.

> However, it seems that if one of the fields is binary (for instance, files), they don't copy the file to the curl command. I suppose that they don't want random bytes in the clipboard.

Ohh i see, that's interesting, in that case i think all i could do is some kind of parsing for those urlencoded values and put there in that format if they come that way and just ignore the files if that's the actual body.

> There is a method in the Gtk.Application class called get_active_window(). Using CarteroApplication::get(), you should be able to retrieve the CarteroApplication instance and then it should let you use the get_active_window() method.

> I am looking for examples and there is one in app.rs here, but a cast is required from GtkWindow to CarteroWindow.

Ok that's certainly interesting, i could try doing it that way! thank you for the example aswell 👍🏻

**Comment by AlphaTechnolog** [comment]
I think with these last changes it's ready for code review.

File: data/meson.build [hunk 1]
```diff
@@ -25,7 +25,9 @@ blueprint_files = [
   'ui/key_value_row.blp',
```
Comment (danirod) [inline data/meson.build:25]: sort

File: data/ui/code_export_pane.blp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        Gtk.Box {
+          orientation: horizontal;
+          spacing: 6;
+
+          Image {
+            icon-name: "edit-copy-symbolic";
+          }
+
+          Label {
+            label: "Copy";
```
Comment (danirod) [inline data/ui/code_export_pane.blp]: _

File: src/widgets/export_tab/code.rs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+                clipboard.set_text(contents.as_ref());
+
+                let app = CarteroApplication::get();
+
+                let window = match app.active_window() {
+                    Some(window) => window.downcast::<CarteroWindow>().unwrap(),
+                    None => CarteroWindow::new(&app),
+                };
+
+                window.toast_message("Content copied to the clipboard");
```
Comment (danirod) [inline src/widgets/export_tab/code.rs]: _

File: src/widgets/export_tab/service.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        Self { endpoint_data }
+    }
+
+    pub fn generate(&self) -> Result<String, CarteroError> {
+        let bound_request = BoundRequest::try_from(self.endpoint_data.clone())?;
+        let mut command = "curl".to_string();
+
+        command.push_str(&{
+            let method_str: String = bound_request.method.into();
+            format!(" -X {} {}", method_str, bound_request.url)
```
Comment (danirod) [inline src/widgets/export_tab/service.rs]: comillas

**Review by danirod — APPROVED** [review]

**Pull request merged**